src/socket.rs

```diff
@@ -73,99 +68,51 @@ pub fn set_transparent(_: &TcpListener) -> io::Result<()> {
 #[cfg(target_os = "linux")]
 #[allow(unsafe_code)]
 mod linux {
-    use std::net::{Ipv4Addr, Ipv6Addr, SocketAddr, SocketAddrV4, SocketAddrV6};
-    use std::os::unix::io::RawFd;
-    use std::{io, mem};
-
-    use log::warn;
-
-    pub unsafe fn so_original_dst(fd: RawFd) -> io::Result<SocketAddr> {
-        let mut sockaddr: libc::sockaddr_storage = mem::zeroed();
-        let mut socklen: libc::socklen_t = mem::size_of::<libc::sockaddr_storage>() as u32;
-
-        let ret = libc::getsockopt(
-            fd,
-            libc::SOL_IP,
-            libc::SO_ORIGINAL_DST,
-            &mut sockaddr as *mut _ as *mut _,
-            &mut socklen as *mut _ as *mut _,
-        );
-        if ret != 0 {
-            let e = io::Error::last_os_error();
-            warn!("failed to read SO_ORIGINAL_DST: {:?}", e);
-            return Err(e);
+    use std::os::unix::io::AsRawFd;
+
+    use socket2::{SockAddr, SockRef};
+    use tokio::io;
+
+    // Replace with socket2's version once there is a release that contains
+    // https://github.com/rust-lang/socket2/pull/360
```

Comment:
this is merged. I guess we may need to wait for a release though?

Reply:
Yeah, especially because `master` started development for `v0.5.x` and they didn't make any releases of that yet.